UCP: improve UT coverage of package `planner`, part 1

hsqlu opened the pull request:
Signed-off-by: Qiannan Lyu <lvqiannan@gmail.com>

UCP: #15681 
<!-- Thank you for contributing to TiDB!

PR Title Format:
1. pkg [, pkg2, pkg3]: what's changed
2. *: what's changed

-->

### What problem does this PR solve?

Issue Number: close #15681 
Problem Summary:

### What is changed and how it works?

What's Changed: Added more test cases

How it Works:

Improved code test coverage.


### Check List <!--REMOVE the items that are not applicable-->

Tests <!-- At least one of them must be included. -->

- Unit test

hsqlu commented:
@winoros  Can you have a look please?

zz-jason commented:
Could you post the coverage improvement here?

zz-jason commented on session/session_test.go:
```diff
@@ -3070,6 +3079,8 @@ func (s *testSessionSuite2) TestStmtHints(c *C) {
 	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
 	c.Assert(tk.Se.GetSessionVars().StmtCtx.MemTracker.CheckBytesLimit(val), IsTrue)
+	tk.MustExec("select /*+ MEMORY_QUOTA(-1 GB) */ 1;")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
```
please also check the warning message.

hsqlu replied:
The `MEMORY_QUOTA` value cannot be negative, so L340 cannot be accessed?

```
"MEMORY_QUOTA" '(' QueryBlockOpt hintIntLit UnitOfBytes ')'
	{
		maxValue := uint64(math.MaxInt64) / $5
		if $4 <= maxValue {
			$$ = &ast.TableOptimizerHint{
				HintName: model.NewCIStr($1),
				HintData: int64($4 * $5),
				QBName:   model.NewCIStr($3),
			}
		} else {
			yylex.AppendError(ErrWarnMemoryQuotaOverflow.GenWithStackByArgs(math.MaxInt64))
			parser.lastErrorAsWarn()
			$$ = nil
		}
	}
```

zz-jason commented on session/session_test.go:
```diff
@@ -2945,12 +2945,21 @@ func (s *testSessionSuite2) TestTxnGoString(c *C) {
 	tk.MustExec("insert into MaxExecTime (id,name,age) values (1,'john',18),(2,'lary',19),(3,'lily',18);")
 
+	tk.MustQuery("select /*+ MAX_EXECUTION_TIME(1000) MAX_EXECUTION_TIME(500) */ * FROM MaxExecTime;")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.GetWarnings(), HasLen, 1)
```
ditto

hsqlu commented:
> Could you post the coverage improvement here?

Improvement can't be shown from the coverage percentage :(

SunRunAway commented on bindinfo/bind_test.go:
```diff
@@ -432,6 +432,26 @@ func (s *testSuite) TestBindingSymbolList(c *C) {
+
+	tk.MustQuery("select a, b from t where a = 3 limit 1, 100")
+	c.Assert(tk.Se.GetSessionVars().StmtCtx.IndexNames[0], Equals, "t:ia")
+	c.Assert(tk.MustUseIndex("select a, b from t where a = 3 limit 1, 100", "ia(a)"), IsTrue)
```
Better to add one more test about index `b`.

hsqlu commented:
@winoros @SunRunAway PTAL

zz-jason commented:
> > Could you post the coverage improvement here?
> 
> Improvement can't be shown from the coverage percentage :(

You can use the following method to generate the UT coverage and upload the coverage file to the PR description:

```sh
cd planner
go test -coverprofile=coverage.out # runs the unit test, writes the coverage profile to coverage.out
go tool cover -html=coverage.out   # displays the coverage report in your browser
```

hsqlu commented:
@zz-jason PTAL.

The command you mentioned doesn't work.
```
➜  planner git:(master) go test -coverprofile=coverage.out
?       github.com/pingcap/tidb/planner [no test files]
```

I changed it to below. Optimize() was improved from 92.1% to 96.8% by `bind_test.go`

```
➜  bindinfo git:(master) go test -coverpkg=../planner -coverprofile=coverage.out ./bind_test.go
ok      command-line-arguments  1.289s  coverage: 73.3% of statements in ../planner
➜  bindinfo git:(master) go tool cover -func=coverage.out                                      
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        92.1%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        85.2%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   95.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         75.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               85.7%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 43.1%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    73.3%
➜  bindinfo git:(master) gco UCP-improve-UT-planner-1    
Switched to branch 'UCP-improve-UT-planner-1'
➜  bindinfo git:(UCP-improve-UT-planner-1) go test -coverpkg=../planner -coverprofile=coverage.out ./bind_test.go
ok      command-line-arguments  1.246s  coverage: 74.8% of statements in ../planner
➜  bindinfo git:(UCP-improve-UT-planner-1) go tool cover -func=coverage.out                                      
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        96.8%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        85.2%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   95.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         75.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               85.7%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 43.1%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    74.8%
```


And coverage for handleStmtHints was improved from 93.1% to 96.6%.

```
➜  session git:(master) go tool cover -func=coverage.out                             
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        36.5%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        96.3%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   80.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         0.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               0.0%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 93.1%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    64.8%
```

```
➜  session git:(UCP-improve-UT-planner-1) go tool cover -func=coverage.out                             
github.com/pingcap/tidb/planner/optimize.go:41:         Optimize                        36.5%
github.com/pingcap/tidb/planner/optimize.go:143:        optimize                        96.3%
github.com/pingcap/tidb/planner/optimize.go:193:        extractSelectAndNormalizeDigest 100.0%
github.com/pingcap/tidb/planner/optimize.go:213:        getBindRecord                   80.0%
github.com/pingcap/tidb/planner/optimize.go:244:        handleInvalidBindRecord         0.0%
github.com/pingcap/tidb/planner/optimize.go:258:        handleEvolveTasks               0.0%
github.com/pingcap/tidb/planner/optimize.go:277:        useMaxTS                        100.0%
github.com/pingcap/tidb/planner/optimize.go:289:        OptimizeExecStmt                0.0%
github.com/pingcap/tidb/planner/optimize.go:305:        handleStmtHints                 96.6%
github.com/pingcap/tidb/planner/optimize.go:397:        init                            100.0%
total:                                                  (statements)                    65.7%
```
